src/api/schema/log_events/mod.rs

```diff
@@ -0,0 +1,170 @@
+        create_log_events_stream(watch_rx, component_names, interval as u64, limit as usize)
+    }
+}
+
+/// Creates a log events stream based on component names, and a provided interval. Will emit
+/// control messages that bubble up the application if the sink goes away. The stream contains
+/// all matching events; filtering should be done at the caller level.
+fn create_log_events_stream(
+    watch_rx: WatchRx,
+    component_names: Vec<String>,
+    interval: u64,
+    limit: usize,
+) -> impl Stream<Item = Vec<LogEventPayload>> {
+    // Channel for receiving individual tap payloads. Since we can process at most `limit` per
+    // interval, this is capped to the same value.
+    let (tap_tx, mut tap_rx) = mpsc::channel(limit);
+
+    // The resulting vector of `LogEventPayload` sent ot the client. Only one result set will be streamed
+    // back to the client at a time. This value is set higher than `1` to prevent blocking the event
+    // pipeline on slower client connections, but low enough to apply a modest cap on mem usage.
+    let (mut log_tx, log_rx) = mpsc::channel::<Vec<LogEventPayload>>(10);
+
+    tokio::spawn(async move {
+        // Create a tap sink. When this drops out of scope, clean up will be performed on the
+        // event handlers and topology observation that the tap sink provides.
+        let _tap_sink = TapSink::new(watch_rx, tap_tx, &component_names);
+
+        // A tick interval to represent when to 'cut' the results back to the client.
+        let mut interval = time::interval(time::Duration::from_millis(interval));
+
+        // Collect a vector of results, with a capacity of `limit`. As new `LogEvent`s come in,
+        // they will be sampled and added to results.
+        let mut results = Vec::<LogEventPayload>::with_capacity(limit);
+
+        // Random number generator to allow for sampling. Speed trumps cryptographic security here.
+        // The RNG must be Send + Sync to use with the `select!` loop below.
+        let mut rng = SmallRng::from_entropy();
+
+        // Keep a count of the batch size, which will be used as a seed for random eviction
+        // per the sampling strategy used below.
+        let mut batch = 0;
+
+        loop {
+            select! {
+                // Process `TapPayload`s. A tap payload could contain a `LogEvent` or a notification.
+                // Notifications are emitted immediately; log events buffer until the next `interval`.
+                Some(tap) = tap_rx.next() => {
+                    let tap = tap.into();
+
+                    // Emit notifications immediately; these don't count as a 'batch'.
+                    if let LogEventPayload::Notification(_) = tap {
+                        // If an error occurs when sending, the subscription has likely gone
+                        // away. Break the loop to terminate the thread.
+                        if log_tx.send(vec![tap]).await.is_err() {
```

Comment:
Could we log this error for debugging purposes?

Reply:
Added in 19aa7cd7e1bd9ee4a91bd8140a5e1448e9f8a7b1.

Reply:
I should note here, too, that I'd have preferred to decorate this debug message with a little more context, but since `tap` is already consumed by the failed send, that would seem to make cloning and/or holding onto references more expensive for what should only happen as part of a client routinely going away... so I kept that bare.

Reply:
👍 thanks for adding the debug statements! Is the client going away the only condition under which the `send` would fail?

I was hoping to add the errors themselves to the log messages too. For example:

```
if let Err(err) = tx.send(TapPayload::LogEvent(component_name.clone(), ev)).await {
  debug!(message = "Couldn't send log event.", component_name = ?component_name, ?err);
  break;
}
```

for https://github.com/timberio/vector/commit/19aa7cd7e1bd9ee4a91bd8140a5e1448e9f8a7b1#diff-7c333535b5def3ed317505d0c5e448be4e1d3da1da9e1a71c85362a12cb3b1e4R106

Reply:
Yes, the only reason a `send` would fail here, is due to the channel being closed. And the only reason a channel could close would be because its receiver has gone away, which is what happens when the event handler goes away (because topology has changed, making it redundant) or the subscription stream is terminated (either explicitly, via a `stop` frame, or the WS terminating.)

This is encoded in the `Err(SendError<T>)` this function returned-- the static string "channel closed" would always wind up as the error message.

I _have_ included this elsewhere, so even though the string isn't very interesting and won't encode the specifics of what happened to the receiver, it would still give it parity by including it. Which I'll do now.

Reply:
Aha, gotcha, thanks for the explanation. I concede that the error doesn't add too much more context, but it would be nice for parity.